<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Asynchronous JavaScript</title>
  </head>

  <body>
    <h1>Asynchronous JavaScript</h1>

    <script>
      const second = () => {
        setTimeout(() => {
          // asynchronous callback
          console.log("Asynchronous hello");
        }, 2000)
      }

      // Synchronous javascript
      // each line code executes one after another as they are written
      const first = () => {
        console.log("First");
        second();
        console.log("The end")
      }

      // first();

      // AJAX example
      // Callback hell example avoid by using promises
      function getRecipe() {
        // simulate response from server
        setTimeout(() => {
          const recipeID = [543, 445, 532, 534, 534];
          console.log(recipeID);

          // retrieve recipe for on of the returned IDs
          setTimeout((id) => {
            const recipe = {
              title: "Fresh tomato pasta",
              publisher: "Jonas"
            };
            console.log(`${id}: ${recipe.title}`);

            setTimeout(publisher => {
              const recipe = {
                title: "Margarita pizza",
                publisher: publisher
              };
              console.log(recipe);
            }, 1500, recipe.publisher);

          }, 1500, recipeID[2]);

        }, 1500);

      }

      // getRecipe();

      // Using ES6 promises to deal with asynchronous js
      const getIDs = new Promise((resolve, reject) => {
        // executor function
        setTimeout(() => {
          // successful promise
          resolve([543, 445, 532, 534, 534]);

          // failed promise
          // reject("failed promise");
        }, 1500);
      });

      const getARecipe = recID => {
        return new Promise((resolve, reject) => {
          setTimeout(ID => {
            const recipe = {
              title: "Fresh tomato pasta",
              publisher: "Jonas"
            };
            resolve(`${ID}: ${recipe.title}`);
          }, 1500, recID);
        });
      };

      const getRelated = publisher => {
        return new Promise((resolve, reject) => {
          setTimeout(pub => {
            const recipe = {
              title: "Margarita pizza",
              publisher: publisher
            };
            resolve(`${pub}: ${recipe.title}`);
          }, 1500, publisher);
        });
      };

      getIDs
      .then(IDs => {
        // handle the successfully fullfilled promise
        console.log(IDs);
        return getARecipe(IDs[2]);  // returns a promise
      }).
      then(recipe => {
        console.log(recipe);
        return getRelated("Jonas");
      }).
      then(recipe => {
        console.log(recipe);
      })
      .catch(error => {
        // handle the failed promise
        // test this by commenting the resolve response in the executor func
        console.log(error);
      });
    </script>
  </body>
</html>
